Added Domino-Simulation-file for Review

SanjayMarreddi opened the pull request:
#309

pep8speaks commented:
Hello @SanjayMarreddi! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`docs/examples/physics_using_pybullet/viz_domino.py`](https://github.com/fury-gl/fury/blob/4388749dae19a327bb58a39f84945c6884733507/docs/examples/physics_using_pybullet/viz_domino.py):

> [Line 44:25](https://github.com/fury-gl/fury/blob/4388749dae19a327bb58a39f84945c6884733507/docs/examples/physics_using_pybullet/viz_domino.py#L44): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 72:40](https://github.com/fury-gl/fury/blob/4388749dae19a327bb58a39f84945c6884733507/docs/examples/physics_using_pybullet/viz_domino.py#L72): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 72:42](https://github.com/fury-gl/fury/blob/4388749dae19a327bb58a39f84945c6884733507/docs/examples/physics_using_pybullet/viz_domino.py#L72): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 76:39](https://github.com/fury-gl/fury/blob/4388749dae19a327bb58a39f84945c6884733507/docs/examples/physics_using_pybullet/viz_domino.py#L76): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 77:39](https://github.com/fury-gl/fury/blob/4388749dae19a327bb58a39f84945c6884733507/docs/examples/physics_using_pybullet/viz_domino.py#L77): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 78:39](https://github.com/fury-gl/fury/blob/4388749dae19a327bb58a39f84945c6884733507/docs/examples/physics_using_pybullet/viz_domino.py#L78): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 137:1](https://github.com/fury-gl/fury/blob/4388749dae19a327bb58a39f84945c6884733507/docs/examples/physics_using_pybullet/viz_domino.py#L137): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 213:80](https://github.com/fury-gl/fury/blob/4388749dae19a327bb58a39f84945c6884733507/docs/examples/physics_using_pybullet/viz_domino.py#L213): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (89 > 79 characters)
> [Line 238:1](https://github.com/fury-gl/fury/blob/4388749dae19a327bb58a39f84945c6884733507/docs/examples/physics_using_pybullet/viz_domino.py#L238): [W391](https://duckduckgo.com/?q=pep8%20W391) blank line at end of file

SanjayMarreddi commented:
@Nibba2018 Could you please give a Feedback on why It is giving Wrong Output.
Thanks !

Nibba2018 commented on docs/examples/physics_using_pybullet/viz_domino.py:
```diff
@@ -0,0 +1,238 @@
+n_dominoes = 10
+centers_list = np.zeros((n_dominoes, 3))
+
+# Adding the dominoes
+for i in range(n_dominoes):
+    center_pos = np.array([(i*0.99)-2.8,1,0])
+    brick_centers[i] = center_pos
+    brick_orns[i] = np.array([0, 0, 0, 1])
+    bricks[i] = p.createMultiBody(baseMass=brick_mass,
+                                      baseCollisionShapeIndex=brick_coll,
+                                      basePosition=center_pos,
+                                      baseOrientation=brick_orns[i])
+    p.changeDynamics(bricks[i], -1, lateralFriction=0.1, restitution=0.1)
```
Your domino position calculation is a bit wrong. Your dominoes are being rendered inside the BASE plate and that why you can observe a weird bump during the start of the simulation.

SanjayMarreddi replied:
@Nibba2018 Yeah This is the place I am facing problem.. Now I have changed the Position and There is no bump in the start of the Simulation. **But I have a doubt why are the Domino's going below below the Base during Simulation?** I have kept center of the Base as `(0,0,0)`. Since Size of the bricks is `np.array([2,0.8, 0.2])` I have kept the y coordinate of the Bricks as 0.4 now like this -
```
for i in range(n_dominoes):
    center_pos = np.array([(i*0.99)-3.3,0.4,1.5])
```

**So Now when I am running the Code, Initially the bricks are lying exactly with their bottom ends on the base, But when Simulation is proceeding, The bottom ends of Bricks are going below the base !!!**

Please check the file now(I made commit in the same PR) and help me resolve this 

Thanks !

Nibba2018 replied:
![image](https://user-images.githubusercontent.com/29832615/97733577-0c946800-1afe-11eb-9df6-08bf00bdf1c8.png)
@SanjayMarreddi, this is how your dominoes are being initialised in pybullet.

Nibba2018 replied:
I think you are getting your orientation wrong.

SanjayMarreddi replied:
@Nibba2018 
I have corrected the Orientation Problem. **Now The Simulation is working Correctly in Pybullet World (GUI option) !**
But in the FURY World, The Dominoes are coming in Horizontal position because of which the Simulation in FURY world is not completely Correct. Where as in the Pybullet world they are fine and in Vertical Position.
I have committed my changes in the same PR.
Could you please tell me how to approach in resolving this !!

Thanks !

Nibba2018 commented on docs/examples/physics_using_pybullet/viz_domino.py:
```diff
@@ -0,0 +1,238 @@
+import itertools
+import pybullet as p
+
+# Instantiate Pybullet client.
+client = p.connect(p.DIRECT)
+
```
If you are not sure whats going on then you can do `client = p.connect(p.GUI)` to understand render in pybullet world.

Nibba2018 commented on docs/examples/physics_using_pybullet/viz_domino.py:
```diff
@@ -0,0 +1,238 @@
+# orientation.
+
+
+def sync_actor(actor, multibody):
+    pos, orn = p.getBasePositionAndOrientation(multibody)
+    actor.SetPosition(*pos)
+    orn_deg = np.degrees(p.getEulerFromQuaternion(orn))
+    actor.SetOrientation(*orn_deg)
+
```
This method is to update the position and orientation of the entire actor. This is actually not necessary. You can render the first domino without any issues in `sync_bricks`.

Nibba2018 commented on docs/examples/physics_using_pybullet/viz_domino.py:
```diff
@@ -0,0 +1,238 @@
+                             flags=p.WORLD_FRAME)
+        apply_force = False
+
+    # Set position and orientation of the first Brick.
+    sync_actor(brick_actor, bricks[0])
```
All your dominoes are rendered by this single actor called `brick_actor`. So whatever modifications you make to this actor will be reflected on all the dominoes.
Thus, in this case when you pass `brick_actor`, `bricks` to `sync_actor` you are basically asking fury to modify all the dominoes with the position and orientation of the first domino, which is incorrect.

That's why you can observe that all of your dominoes are following the same trajectory.

Nibba2018 commented on docs/examples/physics_using_pybullet/viz_domino.py:
```diff
@@ -0,0 +1,238 @@
+* Update the position and orientation.
+"""
+
+def sync_brick(object_index, multibody):
+    pos, orn = p.getBasePositionAndOrientation(multibody)
+
+    rot_mat = np.reshape(
+        p.getMatrixFromQuaternion(
+            p.getDifferenceQuaternion(orn, brick_orns[object_index])),
+        (3, 3))
+
+    vertices[object_index * sec: object_index * sec + sec] = \
+        (vertices[object_index * sec: object_index * sec + sec] -
+         brick_centers[object_index]) @ rot_mat + pos
+
+    brick_centers[object_index] = pos
+    brick_orns[object_index] = orn
```
This method is responsible to update the position and orientation of the objects. So if according to physics the domino is supposed to be stationary then it will be stationary. So you can safely pass the first domino here.

Nibba2018 reviewed (commented):
@SanjayMarreddi I have added some comments below.

Nibba2018 commented on docs/examples/physics_using_pybullet/viz_domino.py:
```diff
@@ -0,0 +1,224 @@
+for i in range(n_dominoes):
+    center_pos = np.array([(i*0.99)-5,0.4,1])
+    brick_centers[i] = center_pos
+    if i == 0 : 
+        brick_orns[0] = np.array([-0.6 , -0.6, 1, 1])
+    else:
+        brick_orns[i] = np.array([1 ,1, 1, 1])
```
@SanjayMarreddi This is the part of the code that is ruining the positioning of the bricks. These values are in `quaternion`. Remove this part and the simulation should work fine.

SanjayMarreddi replied:
@Nibba2018 
I kept the Orientation of the First Brick in a Slanted Position willingly. So that the Brick will yield to Domino effect even though we apply Force at COM. (We can even change this scenario by keeping the Brick Parallel only and apply force above COM as you said)

**But my Doubt is the Following:-
Currently, With the code (with out removing above 73-76 lines), The Simulation is working Correctly in the PyBullet World(GUI). But in the FURY World, The Dominoes are coming in "Horizontal position" because of which the Simulation in FURY world is not completely Correct.** 

Please find the Images of the Same:-

1) In the Pybullet :
![Screenshot (191)](https://user-images.githubusercontent.com/57671048/97831526-4f209500-1cf6-11eb-9a87-610b605d4852.png)

2) In the Fury:
![Screenshot (192)](https://user-images.githubusercontent.com/57671048/97831560-695a7300-1cf6-11eb-8a94-c0ffb0a9a246.png)

Nibba2018 replied:
@SanjayMarreddi In these lines you are explicitly assigning values to `brick_orns` which is only used by pybullet and not by fury. If you need to make changes to the fury world then you have to update `brick_directions` as well. Now as you are only updating `brick_orns` instead of updating both, it means that your simulation in fury will have a negative offset(in orientation) compared to the one in pybullet.

Just remove those line and you will notice that pybullet and fury are both in sync. Pybullet and Fury uses different methods to represent orientations and we haven't figured out a way to map those together. So, we let `sync_bricks` handle the orientation situation during runtime.

SanjayMarreddi replied:
@Nibba2018 
I understood your explanation. But the problem is If I remove those 4 lines, I could not see any thing in both the FURY and PYBULLET WORLD !

So I am trying to change `brick_directions`

Nibba2018 replied:
@SanjayMarreddi right for that you need to set `brick_directions[:] = np.array([1.57, 0, 0])` and `brick_orns[i] = np.array([0, 0, 0, 1])`

SanjayMarreddi replied:
@Nibba2018  Basically I started from this only (In the first Commit of PR).
Since it is Initializing the things in a Wrong way, I started changing the Orientation to   
```
brick_orns[i] = np.array([1 ,1, 1, 1])
``` 
to make the Bricks Vertical in Pybullet (GUI).

In order to apply force on the COM on 1st Brick, I tilted the Orientation of 1st Brick and finally changed to -

```    
if i == 0 : 
        brick_orns[0] = np.array([-0.6 , -0.6, 1, 1])
    else:
        brick_orns[i] = np.array([1 ,1, 1, 1])
```
So Now it is working Fine in GUI

and then I tried Multiple Directions for FURY world to fix orientation problem in FURY like -
```
brick_directions[:] = np.array([1.57, 0, 0])
brick_directions[:] = np.array([0, 1, 1]) # This is giving better result
```

The result corresponding to `brick_directions[:] = np.array([0, 1, 1])` is as follows:-

![Screenshot (197)](https://user-images.githubusercontent.com/57671048/97846299-cca6ce00-1d13-11eb-93a7-cabe540cefe8.png)

**The only thing left is to make this similar to the Pybullet by changing `brick_directions[:] = np.array([0, 1, 1])` to a better value. I have been trying soo hard for this.. Please suggest the direction.**

Thanks!

Nibba2018 replied:
@SanjayMarreddi Is it really necessary to tilt the domino? I mean you are already applying a force in `timer_callback`. Trying to tilt it will complicate things. I have made a PR on your repo kindly check.

SanjayMarreddi replied:
@Nibba2018 

Okay Got it.. Thanks .!

So finally, The entire Problem between the Syncing of FURY and PYBULLET occurred because of Including the Tilt in the First domino. So to rectify this we kept it parallel only and applied force at a point above COM..

Nibba2018 commented on docs/examples/physics_using_pybullet/viz_domino.py:
```diff
@@ -0,0 +1,224 @@
+    if apply_force:
+        # Apply the force.
+        p.applyExternalForce(bricks[0], -1,
+                             forceObj=[400, 0, 0],
+                             posObj=brick1_pos,
+                             flags=p.WORLD_FRAME)
```
Here, you are applying the force at the center of mass of the domino so it will never rotate in this situation. You need to apply the force above the center of mass.

Nibba2018 commented:
![Peek 2020-11-01 23-48](https://user-images.githubusercontent.com/29832615/97810846-d04b3e00-1c9c-11eb-8c57-4936f54154de.gif)

SanjayMarreddi commented:
@Nibba2018 
I have made all the Final Changes. Could you please review it.

Also Could you tell me more about `TravisCI - Pull Request` automated Test.
I am failing in that test. To make my PR merged, I need to pass all the tests I think?

Thanks !

skoudoro commented:
Good progress here!  do not forget to address all pep8 issues reported here[https://github.com/fury-gl/fury/pull/334#issuecomment-718590194] @SanjayMarreddi.

Also, you will need to document and explain a bit more your tutorial. It is missing a title and explanation. Please, look at the other tutorials to get an idea of the format.

No worries about Travis, the error does not seem to be related to your PR. I will test it a bit later.

SanjayMarreddi commented:
@skoudoro 
Sure ! I will address all the pep8 issues Reported.
Also I will document and explain the Tutorial completely by Tomorrow night, Hope that is fine.

Thanks !

Nibba2018 commented:
@SanjayMarreddi  Also make sure to change your variable names from `brick_*` to `domino_*`.

SanjayMarreddi commented:
@Nibba2018 Sure I will do that as well..

SanjayMarreddi commented:
@Nibba2018 @skoudoro 
Please review the PR Now.
I resolved all the PEP Issues, Added the Title , Improved Documentation and Explanation of Tutorial, Changed the Variable Names too..

Thanks !

Nibba2018 commented on docs/examples/physics_using_pybullet/viz_domino.py:
```diff
@@ -0,0 +1,228 @@
+"""
+=====================
+Domino Physics Simulation
+=====================
```
The `====` must cover the entire text.

Nibba2018 commented on docs/examples/physics_using_pybullet/viz_domino.py:
```diff
@@ -0,0 +1,228 @@
+under Domino Effect.
+
+"""
+###############################################################################
```
I don't think this line is necessary :thinking: . As the doc-string itself will be considered as a different section.

Nibba2018 commented on docs/examples/physics_using_pybullet/viz_domino.py:
```diff
@@ -0,0 +1,228 @@
+                       directions=[0, 0, 0],
+                       scales=base_size,
+                       colors=base_color)
+
+base_coll = p.createCollisionShape(p.GEOM_BOX,
+                                   halfExtents=base_size / 2)
+
+# half of the actual size.
```
This comment should be above line 38.

Nibba2018 commented on docs/examples/physics_using_pybullet/viz_domino.py:
```diff
@@ -0,0 +1,228 @@
+    # Simulate a step.
+    p.stepSimulation()
+
+    # Exit after 2000 steps of simulation.
+    if cnt == 2000:
+        showm.exit()
```
Reduce `2000` to `300` steps.

Nibba2018 commented on docs/examples/physics_using_pybullet/viz_domino.py:
```diff
@@ -0,0 +1,228 @@
+if interactive:
+    showm.start()
+
+window.record(scene, out_path="domnio_simulation.png", size=(900, 768))
+###############################################################################
```
The `#####` line isn't necessary. also set `out_path="viz_domino.png"` to make it same as the other examples.

Nibba2018 reviewed (commented):
Great Job! @SanjayMarreddi . The simulation was very well made. I have added some comments below.

SanjayMarreddi commented:
@Nibba2018 
Made the Above Modifications. Please Review it !

Thanks!

Nibba2018 reviewed (approved):
LGTM! :rocket: . I would like to request @skoudoro to have a look.

skoudoro commented:
Thank you for the work @SanjayMarreddi  and thanks for the review @Nibba2018!

I will look into it by the end of the week and keep you updated

SanjayMarreddi commented:
@skoudoro Kindly look into this PR & provide feedback.. Will be waiting for your update. Thanks !!

skoudoro commented on docs/examples/physics_using_pybullet/viz_domino.py:
```diff
@@ -0,0 +1,226 @@
+domino_mass = 0.5
+domino_size = np.array([0.1, 1, 2])
+
+number_of_dominoes = 10
```
If I change the number of dominoes,  it will be good if you could update the size of the base. Otherwise, it looks strange

skoudoro replied:
you can check the simulation when you change the value to 20.

skoudoro commented on docs/examples/physics_using_pybullet/viz_domino.py:
```diff
@@ -0,0 +1,226 @@
+sec = np.int(num_vertices / num_objects)
+
+###############################################################################
+# ==============
```
`===` should go until the end of `Dominoes`, otherwise there is a error on doc generation

skoudoro commented on docs/examples/physics_using_pybullet/viz_domino.py:
```diff
@@ -0,0 +1,226 @@
+###############################################################################
+# ==============
+# Syncing Dominoes
+# ==============
```
same comment as above

skoudoro reviewed (commented):
Hi @SanjayMarreddi,

Overall, it looks good to me, Nice job! See below for some comments but I think it is ready to go soon. Sorry for the late review.

SanjayMarreddi commented:
@skoudoro 
I made all the above Modifications suggested by you. Could you please re-review it now ?
Thanks !

skoudoro commented:
thank you for the update @SanjayMarreddi. This first version looks good. merging
